File: Ahorn/entities/FallenLeafPile.jl
module FallenLeafPile

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/FallenLeafPile" FallenLeafPile(x::Integer, y::Integer, collectibleType::String="", detectionRange::Integer=40, enemyType::String="MaggyHelper/RuinsSentinel", hiddenContent::String="Nothing")

const placements = Ahorn.PlacementDict(
    "normal" => Ahorn.EntityPlacement(FallenLeafPile),
    "spikes" => Ahorn.EntityPlacement(FallenLeafPile),
    "enemy" => Ahorn.EntityPlacement(FallenLeafPile),
    "collectible" => Ahorn.EntityPlacement(FallenLeafPile)
)

function Ahorn.selection(entity::FallenLeafPile)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FallenLeafPile, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/fallenLeafPile.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local fallenLeafPile = {}

fallenLeafPile.name = "MaggyHelper/FallenLeafPile"
fallenLeafPile.depth = -50
fallenLeafPile.fieldInformation = {
    hiddenContent = {
        options = {"Nothing", "Spikes", "Enemy", "Collectible", "SecretPath"},
        editable = false
    },
    detectionRange = {
        minimumValue = 0
    }
}
fallenLeafPile.placements = {
    {
        name = "normal",
        data = {
            hiddenContent = "Nothing",
            detectionRange = 40,
            enemyType = "MaggyHelper/RuinsSentinel",
            collectibleType = ""
        }
    },
    {
        name = "spikes",
        data = {
            hiddenContent = "Spikes",
            detectionRange = 40,
            enemyType = "",
            collectibleType = ""
        }
    },
    {
        name = "enemy",
        data = {
            hiddenContent = "Enemy",
            detectionRange = 40,
            enemyType = "MaggyHelper/RuinsSentinel",
            collectibleType = ""
        }
    },
    {
        name = "collectible",
        data = {
            hiddenContent = "Collectible",
            detectionRange = 40,
            enemyType = "",
            collectibleType = ""
        }
    }
}

function fallenLeafPile.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/fallen_leaf_pile/idle", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function fallenLeafPile.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 16, 24, 16)
end

return fallenLeafPile
